File: Problems/Algorithms/199. Binary Tree Right Side View/BinaryTreeSideView.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        
        if (root == null) return result;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int len = queue.size();
            
            for (int i = 0; i < len; i++) {
                TreeNode node = queue.poll();
                
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
                
                if (i == len-1) {
                    result.add(node.val);
                }
            }
        }
        
        return result;
    }
}

File: Problems/Algorithms/199. Binary Tree Right Side View/BinaryTreeSideView.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun rightSideView(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        
        if (root == null) return result
        
        val queue = ArrayDeque<TreeNode?>()
        queue.add(root)
        
        while (!queue.isEmpty()) {
            val len = queue.size
            
            for (i in 0..len-1) {
                val node = queue.removeFirst()
                
                if (node?.left != null) {
                    queue.add(node?.left)
                }
                if (node?.right != null) {
                    queue.add(node?.right)
                }
                
                if (i == len-1) {
                    result.add(node?.`val`!!)
                }
            }
        }
        
        return result
    }
}

